pkg/controller.v1alpha3/experiment/experiment_controller.go

```diff
@@ -193,17 +193,24 @@ func (r *ReconcileExperiment) Reconcile(request reconcile.Request) (reconcile.Re
 	if instance.IsCompleted() {
 		// Check if completed instance is restartable
 		// Experiment is restartable only if it is in succeeded state by reaching max trials
+		// And Resume Policy is LongRunning
 		if util.IsCompletedExperimentRestartable(instance) {
```

Comment:
I think this paragraph could be modified to something below 
```go
	if instance.IsCompleted() {
		needRestart := false
		// Check if completed instance is restartable
		// Experiment is restartable only if it is in succeeded state by reaching max trials
		if util.IsCompletedExperimentRestartable(instance) {
			// Check if max trials is reconfigured
			if (instance.Spec.MaxTrialCount != nil &&
				*instance.Spec.MaxTrialCount != instance.Status.Trials) ||
				(instance.Spec.MaxTrialCount == nil && instance.Status.Trials != 0) {
				msg := "Experiment is restarted"
				instance.MarkExperimentStatusRestarting(util.ExperimentRestartingReason, msg)
				needRestart = true
			}
		}
		// If experiment who doesn't need to restart is completed without running trials, stop reconcile
		if !needRestart && !instance.HasRunningTrials() {
			if instance.Spec.ResumePolicy != experimentsv1alpha3.LongRunning {
				return r.terminateSuggestion(instance)
			}
			return reconcile.Result{}, nil
		}
	}
```
Otherwise, `terminateSuggestion` won't work when corresponding experiment has been completed for reaching max trials.

Reply:
@sperlingxx Why `terminateSuggestion` won't run when experiment finishes with reaching max trials?

I believe, Reconcile loop runs with Succeeded experiment state, after reaching this step: https://github.com/kubeflow/katib/blob/master/pkg/controller.v1alpha3/experiment/util/status_util.go#L182.

I am wondering, do we actually need to check if Experiment has running trials in the controller @gaocegege @johnugeorge :
https://github.com/kubeflow/katib/blob/master/pkg/controller.v1alpha3/experiment/experiment_controller.go#L209 ?
Or we can just `return reconcile.Result{}, nil` without this check if Experiment is Completed?

Reply:
@andreyvelich I think `util.IsCompletedExperimentRestartable` will be true when experiment finishes with reaching max trials. But, for now, `terminateSuggestion` will only be called  when `util.IsCompletedExperimentRestartable` is false.

Reply:
Yeah, but `util.IsCompletedExperimentRestartable` return True only if `ResumePolicy: LongRunning`.
 
https://github.com/kubeflow/katib/pull/1184/files#diff-f0faa0b63a35acff95fe5e9d93d594ffR201

Reply:
@andreyvelich Oh, I got it!